File: programmers/kakao#2/python/KAKAO23.py
from collections import defaultdict
import heapq

def solution(stones, k):
    ans, left, right = 0, 1, k
    counter = defaultdict(int)
    max_heap = []

    for i in range(k):
        counter[stones[i]] += 1
        ans = max(ans, stones[i])
        heapq.heappush(max_heap, -stones[i])

    while right < len(stones):
        counter[stones[left - 1]] -= 1

        while max_heap and counter[-max_heap[0]] == 0:
            heapq.heappop(max_heap)

        counter[stones[right]] += 1
        heapq.heappush(max_heap, -stones[right])

        ans = min(ans, -max_heap[0])

        right += 1
        left += 1
    print(ans)
    return ans

solution([2, 4, 5, 3, 2, 1, 4, 2, 5, 1], 3)

File: 0_4.programmers/kakao#2/java/KAKAO23.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class KAKAO23 {
    public static void main(String[] args) {
        int result = solution(new int[] {2, 4, 5, 3, 2, 1, 4, 2, 5, 1}, 3);
        System.out.println(result);
    }    
    public static int solution(int[] stones, int k) {
        int ans = 0;
        int left = 1, right = k;
        PriorityQueue<Integer> pq = new PriorityQueue<>((o1, o2) -> o2 - o1);
        Map<Integer, Integer> counter = new HashMap<>();

        for (int i=0;i<k;i++) {
            ans = Math.max(ans, stones[i]);
            counter.putIfAbsent(stones[i], 0);
            counter.compute(stones[i], (key, value) -> value + 1);
            pq.add(stones[i]);
        }

        for (;right<stones.length;right++) {
            counter.compute(stones[left-1], (key, value) -> value - 1);

            while (!pq.isEmpty() && counter.get(pq.peek()) == 0)
                pq.poll();
            
            pq.add(stones[right]);
            counter.putIfAbsent(stones[right], 0);
            counter.compute(stones[right], (key, value) -> value + 1);

            ans = Math.min(ans, pq.peek());

            left++;
        }
        
        return ans;
    }
}
/**
 * [2019 카카오 인턴] 징검다리 건너기
 * 해결: X
 */